packages/template-retail-react-app/app/pages/login/index.jsx

```diff
@@ -27,19 +35,57 @@ const Login = () => {
     const form = useForm()
     const location = useLocation()
     const einstein = useEinstein()
-    const {isRegistered} = useCustomerType()
+    const {isRegistered, customerType} = useCustomerType()
+    const customerId = useCustomerId()
     const login = useAuthHelper(AuthHelpers.LoginRegisteredUserB2C)
 
+    /************** merge basket for recurring users ***/
+    const prevAuthType = useRef()
+
+    const {data: baskets} = useCustomerBaskets(
+        {parameters: {customerId}},
+        {enabled: !!customerId && !isServer, keepPreviousData: true}
+    )
+    const mergeBasket = useShopperBasketsMutation('mergeBasket')
+    /*****************/
     const submitForm = async (data) => {
         try {
-            await login.mutateAsync({username: data.email, password: data.password})
+            await login.mutateAsync(
+                {username: data.email, password: data.password},
+                {
+                    onError: (error) => {
+                        const message = /Unauthorized/i.test(error.message)
+                            ? formatMessage(LOGIN_ERROR_MESSAGE)
+                            : formatMessage(API_ERROR_MESSAGE)
+                        form.setError('global', {type: 'manual', message})
+                    }
+                }
+            )
+            const hasBasketItem = baskets?.baskets?.[0]?.productItems?.length > 0
+            // we only want to merge basket when customerType changes from guest to registered
+            const shouldMergeBasket = hasBasketItem && prevAuthType.current === 'guest'
+            if (shouldMergeBasket) {
```

Comment:
How about adding a comment here in this file (and also in the modal file) to make sure that if anyone updates this logic, they should do the same to the other file?

Reply:
@alexvuong now that I think about it, let's extract this duplicate code into its own custom hook. So that we can call it like this:

```js
const {mergeBasketIfNeeded} = useMergeBasket

const submitFrom = async (data) => {
  try {
    await login.mutateAsync(...)
    mergeBasketIfNeeded()
    ...
  }
}
```

And the custom hook would look something like this:

```js
export const useMergeBasket = () => {
    const prevAuthType = useRef()
    useEffect(() => {
        prevAuthType.current = customerType
    }, [customerType])

    const customerId = useCustomerId()
    const {data: baskets} = useCustomerBaskets(
        {parameters: {customerId}},
        {enabled: !!customerId && !isServer, keepPreviousData: true}
    )

    const mergeBasket = useShopperBasketsMutation('mergeBasket')

    const mergeBasketIfNeeded = () => {
        const hasBasketItem = baskets?.baskets?.[0]?.productItems?.length > 0
        // we only want to merge basket when customerType changes from guest to registered
        const shouldMergeBasket = hasBasketItem && prevAuthType.current === 'guest'
        if (shouldMergeBasket) {
            mergeBasket.mutate({
                headers: {
                    // This is not required since the request has no body
                    // but CommerceAPI throws a '419 - Unsupported Media Type' error if this header is removed.
                    'Content-Type': 'application/json'
                },
                parameters: {
                    createDestinationBasket: true
                }
            })
        }
    }

    return {
        mergeBasketIfNeeded
    }
}
```